File: Python/LinkedList/Palindrome Linked List.py
#Given a linked list , check whether it is palindrome or not

#leetcode: https://leetcode.com/problems/palindrome-linked-list

#python code :

class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        slow, fast=head, head
        pre=None
        if head is None or head.next is None:
            return True
        while fast is not None and fast.next is not None:
            pre=slow
            slow=slow.next
            fast=fast.next.next
        if fast is None:   
            h2=self.reverseList(slow)
        else:
            h2=self.reverseList(slow.next)
        pre.next=h2
        while h2 is not None:
            if head.val != h2.val:
                return False
            head=head.next
            h2=h2.next
        return True
            
        
        
        
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head is None or head.next is None:
            return head
        pnode=None
        cnode=head
        
        while cnode is not None:
            nnode=cnode.next
            cnode.next=pnode
            pnode=cnode
            cnode=nnode
        
        head=pnode
        return head

File: 234-palindrome-linked-list/palindrome-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
        private ListNode reverseList(ListNode head) {
        ListNode prev = null;
        ListNode curr = head;

        while (curr != null) {
            ListNode next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }

        return prev;
    }
    public boolean isPalindrome(ListNode head) {
         if (head == null || head.next == null) {
            return true;
        }

        ListNode slow = head;
        ListNode fast = head;
        ListNode pre = null;

        while (fast != null && fast.next != null) {
            pre = slow;
            slow = slow.next;
            fast = fast.next.next;
        }

        ListNode h2;
        if (fast == null) {         
            h2 = reverseList(slow);
        } else {                   
            h2 = reverseList(slow.next);
        }

        pre.next = h2;

        while (h2 != null) {
            if (head.val != h2.val) {
                return false;
            }
            head = head.next;
            h2 = h2.next;
        }

        return true;
    }
}